pkg/metrics/event.go

```diff
@@ -175,3 +183,144 @@ func (events Events) SplitPayload(times int) ([]marshaler.Marshaler, error) {
 	}
 	return splitPayloads, nil
 }
+
+//// The following methods implement the StreamJSONMarshaler interface
+//// for support of the enable_events_stream_payload_serialization option.
+
+// Initialize the data for serialization. Call once before any other methods.
+func (events Events) Initialize() error {
+
+	// Sort because events are aggregated by SourceTypeName. See WriteItem.
+	sort.Slice(events, func(i, j int) bool { return events[i].SourceTypeName < events[j].SourceTypeName })
+	return nil
+}
+
+// WriteHeader writes the payload header for this type
+func (events Events) WriteHeader(stream *jsoniter.Stream) error {
+	stream.WriteObjectStart()
+	stream.WriteObjectField(apiKeyJSONField)
+	stream.WriteString("")
+	stream.WriteMore()
+
+	stream.WriteObjectField(eventsJSONField)
+	stream.WriteObjectStart()
+
+	return stream.Flush()
+}
+
+// WriteFooter prints the payload footer for this type
+func (events Events) WriteFooter(stream *jsoniter.Stream) error {
+	// We have at least one event and so we need to close the array
+	stream.WriteArrayEnd()
+	return events.writeNoEventFooter(stream)
+}
+
+// WriteLastFooter writes the last footer. Call once after any other methods.
+func (events Events) WriteLastFooter(stream *jsoniter.Stream, itemWrittenCount int) error {
+	// If we do not write anything we should not close the JSON array.
+	if itemWrittenCount == 0 {
+		return events.writeNoEventFooter(stream)
+	}
+
+	return events.WriteFooter(stream)
+}
+
+func (events Events) writeNoEventFooter(stream *jsoniter.Stream) error {
+	stream.WriteObjectEnd()
+	stream.WriteMore()
+
+	hostname, _ := util.GetHostname()
+	stream.WriteObjectField(internalHostnameJSONField)
+	stream.WriteString(hostname)
+
+	stream.WriteObjectEnd()
+	return stream.Flush()
+}
+
+// WriteItem prints the json representation of an item
+func (events Events) WriteItem(stream *jsoniter.Stream, i int, itemIndexInPayload int) error {
+	if i < 0 || i > len(events)-1 {
+		return errors.New("out of range")
+	}
+
+	event := events[i]
+	isFirstInPayLoad := itemIndexInPayload == 0
+	var startNewSourceType bool
+
+	if isFirstInPayLoad {
+		startNewSourceType = true
+	} else if i > 0 && events[i].SourceTypeName != events[i-1].SourceTypeName {
+		// Close previous source type and open a new one
+		stream.WriteArrayEnd()
+		stream.WriteMore()
+		startNewSourceType = true
+	} else {
+		// As SupportJSONSeparatorInsertion returns false, we need the separator between items
+		stream.WriteMore()
+	}
+
+	if startNewSourceType {
+		sourceTypeName := event.SourceTypeName
+		if sourceTypeName == "" {
+			sourceTypeName = defaultSourceType
+		}
+
+		stream.WriteObjectField(sourceTypeName)
+		stream.WriteArrayStart()
+	}
+
+	if err := writeEvent(event, stream); err != nil {
+		return err
+	}
+	return stream.Flush()
+}
+
+// SupportJSONSeparatorInsertion returns true to add JSON separator automatically between two calls of WriteItem, false otherwise.
```

Comment:
I don't find the name `SupportJSONSeparatorInsertion` very explicit for this function. `AddJSONSeparatoraAutomatically` ?